test/test_mahalanobis_mixin.py

```diff
@@ -0,0 +1,203 @@
+                  'lsml',
+                  'mlkr',
+                  'mmc',
+                  'nca',
+                  'rca',
+                  'sdml',
+                  ]
+
+
+@pytest.mark.parametrize('estimator, build_dataset', list_estimators,
+                         ids=ids_estimators)
+def test_score_pairs_pairwise(estimator, build_dataset):
+  # Computing pairwise scores should return an euclidean distance matrix.
+  inputs, labels = build_dataset()
+  X, _ = load_iris(return_X_y=True)
+  n_samples = 20
+  X = X[:n_samples]
+  model = clone(estimator)
+  model.fit(inputs, labels)
+
+  pairwise = model.score_pairs(np.array(list(product(X, X))))\
+      .reshape(n_samples, n_samples)
+
+  check_is_distance_matrix(pairwise)
+
+  # a necessary condition for euclidean distances matrix: (see
+  # https://en.wikipedia.org/wiki/Euclidean_distance_matrix)
+  assert np.linalg.matrix_rank(pairwise**2) <= min(X.shape) + 2
+
+
+@pytest.mark.parametrize('estimator, build_dataset', list_estimators,
+                         ids=ids_estimators)
+def test_score_pairs_toy_example(estimator, build_dataset):
+    # Checks that score_pairs works on a toy example
+    inputs, labels = build_dataset()
+    X, _ = load_iris(return_X_y=True)
+    n_samples = 20
+    X = X[:n_samples]
+    model = clone(estimator)
+    model.fit(inputs, labels)
+    pairs = np.stack([X[:10], X[10:20]], axis=1)
+    embedded_pairs = pairs.dot(model.transformer_.T)
+    distances = np.sqrt(np.sum((embedded_pairs[:, 1] -
+                               embedded_pairs[:, 0])**2,
+                               axis=-1))
+    assert_array_almost_equal(model.score_pairs(pairs), distances)
+
+
+@pytest.mark.parametrize('estimator, build_dataset', list_estimators,
+                         ids=ids_estimators)
+def test_score_pairs_finite(estimator, build_dataset):
+  # tests that the score is finite
+  inputs, labels = build_dataset()
+  model = clone(estimator)
+  model.fit(inputs, labels)
+  X, _ = load_iris(return_X_y=True)
+  pairs = np.array(list(product(X, X)))
+  assert np.isfinite(model.score_pairs(pairs)).all()
+
+
+@pytest.mark.parametrize('estimator, build_dataset', list_estimators,
+                         ids=ids_estimators)
+def test_score_pairs_dim(estimator, build_dataset):
+  # scoring of 3D arrays should return 1D array (several tuples),
+  # and scoring of 2D arrays (one tuple) should return a scalar (0D array).
+  inputs, labels = build_dataset()
+  model = clone(estimator)
+  model.fit(inputs, labels)
+  X, _ = load_iris(return_X_y=True)
+  tuples = np.array(list(product(X, X)))
+  assert model.score_pairs(tuples).shape == (tuples.shape[0],)
+  assert np.isscalar(model.score_pairs(tuples[1]))
+
+
+def check_is_distance_matrix(pairwise):
+  assert (pairwise >= 0).all()  # positivity
+  assert np.array_equal(pairwise, pairwise.T)  # symmetry
+  assert (pairwise.diagonal() == 0).all()  # identity
+  # triangular inequality
+  for i in range(pairwise.shape[1]):
```

Comment:
not crucial but there must be a more efficient way to do this. at least, since you already checked that the matrix is symmetric you do not need to check all of them

Reply:
Agreed, I'll make the `j` loop start from `i+1` instead of 0 (+1 because with positivity, identity and symmetry `d(x,x) <=d(x,y) + d(y,x)` is already tested), this should reduce the complexity from `n^3` to `n^3/2` I think. I could also prevent from testing if `k==i` or `k== j` since I already tested identity, but this would not be so significant, so maybe no need to include this `if` in the code
I don't know if it could be made even more efficient

Reply:
How about this:

```python
rows, cols = pairwise.shape
tol = 1e-3  # seems kinda large, but we can tune this
for i in range(rows):
  for j in range(i, cols):
    direct_dist = pairwise[i, j]
    triangle_dist = np.add.outer(pairwise[i], pairwise[:, j])
    assert (direct_dist <= triangle_dist + tol).all()